User company favourite - save a company to my favourites

lemocla opened the pull request:
# Business Analysis 🟠 Code Institute 🟠 Community Project - Pull Request 🎉🎉

---

👉 _**Please remove this banner, and the below content under each header and replace with your own values, leaving the headers where they are.**_ 👈

---

## Pull Request References:
   - Issue #7 
   - #U05: As User, I would like to be able to save a selected company to my favorites

## PR contains:

PR contains update the following existing files: 
- models > user.py - organisations.py 
- template > list_organisations.html 
- 
PR contains new files as follows:
 - favourites.py 
 - favourites.js  
 
Adding a company to a list of favourites for a user will use an Ajax call to a python function “add_to_favourites”, which in turn will: 
-  update a user with updating the array field “favourites” by adding the company id 
- if the “favourites” field existing - adding the field “favourites” and adding the company id in the array 
- if the field does not exist  Once a company is added to favourite, the page will reload to location 
- so that is the user is browsing a long list of companies or using filters 
- they will not be disturbed in their navigation. 

**models > user.py**
- Update users model with initialisation of a "favourites' field when user is created
- Add static method in users to find a user by username passing the session[_user_id] as variable
- Add static method in User to append a list of favourite companies as follows:
    - If "favourites" field exist, add the company id to the list
    - If "favourites" field does not exist, set the company id in an array - named "favourites" - adding the field the user document.

**organisations.py**
- Update to views “get_organisations()” to add user collection to check if company id is in favourites 
**templates > list_organisations.html** 
- Add conditional content for adding a company to favourites as follows:
     - If company not in favourites, displayed with blank heart icon
     - if company in favourites items, displayed with black filled heart icon  

**favourites.py**
 - New file to handle functionality to add to favourites
 -  Will also be used to handle favourites views and remove from favourites 
 - Add to favourites functionalities include:  
      - Post request received from script.js / organisation.html
    - Retrieve the company ID from the data
    - Check if company has been added to favourites
    - Add company id to favourites field in user collection in MongoDB
    - Return success message
 
**favourites.js** 
- Create event listeners on icon with data-action = add-favourite 
- On click, call a function to make a post request to “/Add_to_favourites "

## Breaking changes
- May impact the user model and code in PR should check if the field “favourites” exists for an existing user and add it to the document if it does not exist.

## Screenshots/Gifs/Media

https://user-images.githubusercontent.com/74382632/160294112-26a48253-7edf-4c17-abf6-c366e62425a5.mov

stefdworschak commented on app/organisations.py:
```diff
@@ -7,18 +7,22 @@
     organisations = list(mongo.db.organisations.find())
     return render_template('organisations/list_organisations.html',
-                           organisations=organisations)
+                           organisations=organisations, user=user)
```
Since you don't actually need the whole user in the template, could you please just pass in `favourites=user.favourites`

stefdworschak commented on app/static/js/favourites.js:
```diff
@@ -0,0 +1,28 @@
+}).catch((error) => {
+    console.log(error)
+})
+}
```
Please add a blank line eof

stefdworschak commented on app/templates/organisations/list_organisations.html:
```diff
@@ -57,3 +69,8 @@ <h1>Organisations list</h1>
+{% block scripts %}
+<!-- block script for favourites scripts -->
+<script src="{{ url_for('static', filename='js/favourites.js')}}"></script>
+{% endblock %}
```
Please add a blank line eof

stefdworschak commented on static/js/favourites.js:
```diff
@@ -0,0 +1,28 @@
+}).catch((error) => {
+    console.log(error)
+})
+}
```
Please add a blank line eof

stefdworschak commented on app/favourites.py:
```diff
@@ -0,0 +1,45 @@
+
+    return jsonify(message)
+
+    
```
Please remove the extra blank line eof

lemocla replied:
extra blank lined removed

stefdworschak reviewed (changes requested):
Looks good to me overall, please just have a look at the few comments I left and we can merge this.

lemocla commented:
Made changes to user parameter so favourites=user.favourites is passed
Changes also reflected in list_organisations.html
Added space at eof in list_organisations.html, favourites.js
Removed space in favourites.py

Made favourites.js jshint compliant 
Made python files favourites.py and organisations.py pep8 compliant

auxfuse reviewed (approved):
Great PR! And thank you for making the changes that @stefdworschak suggested. 🎉

stefdworschak reviewed (approved):
LGTM
